### ISSUE TITLE
Matrix multiplication broadcasting different from NumPy

### ISSUE THREAD

**Issue by tlaundal** [scikit-hep/awkward#521]
This might be a result of me trying to use Awkward for something it isn't really intended for, but it seems like a bug.

NumPy matrix multiplication always works in the lowest two dimensions and will broadcast over higher dimensions. This does not seem to work as expected in Awkward. Take this example:
```py
import numpy as np
import awkward1 as ak

n = np.array([[1, 2], [3, 4]])
nn = np.array([n, n])
a = ak.Array(n)
aa = ak.Array(nn)

print('n @ n', n @ n)
print('nn @ n', nn @ n)
print('a @ a', a @ a)
print('aa @ a', aa @ a)
```
Which would print
```
n @ n [[ 7 10]
 [15 22]]
nn @ n [[[ 7 10]
  [15 22]]
 [[ 7 10]
  [15 22]]]
a @ a [[7, 10], [15, 22]]
Traceback (most recent call last):
  File "/opt/anaconda/envs/fys-stk4155/lib/python3.8/site-packages/IPython/core/interactiveshell.py", line 3417, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-67-4cd9e7253ac9>", line 12, in <module>
    print('aa @ a', aa @ a)
  File "/opt/anaconda/envs/fys-stk4155/lib/python3.8/site-packages/numpy/lib/mixins.py", line 21, in func
    return ufunc(self, other)
  File "/opt/anaconda/envs/fys-stk4155/lib/python3.8/site-packages/awkward1/highlevel.py", line 1369, in __array_ufunc__
    return awkward1._connect._numpy.array_ufunc(ufunc, method, inputs, kwargs)
  File "/opt/anaconda/envs/fys-stk4155/lib/python3.8/site-packages/awkward1/_connect/_numpy.py", line 173, in array_ufunc
    out = awkward1._util.broadcast_and_apply(
  File "/opt/anaconda/envs/fys-stk4155/lib/python3.8/site-packages/awkward1/_util.py", line 996, in broadcast_and_apply
    out = apply(broadcast_pack(inputs, isscalar), 0)
  File "/opt/anaconda/envs/fys-stk4155/lib/python3.8/site-packages/awkward1/_util.py", line 575, in apply
    return apply(nextinputs, depth)
  File "/opt/anaconda/envs/fys-stk4155/lib/python3.8/site-packages/awkward1/_util.py", line 759, in apply
    outcontent = apply(nextinputs, depth + 1)
  File "/opt/anaconda/envs/fys-stk4155/lib/python3.8/site-packages/awkward1/_util.py", line 606, in apply
    return apply(
  File "/opt/anaconda/envs/fys-stk4155/lib/python3.8/site-packages/awkward1/_util.py", line 759, in apply
    outcontent = apply(nextinputs, depth + 1)
  File "/opt/anaconda/envs/fys-stk4155/lib/python3.8/site-packages/awkward1/_util.py", line 606, in apply
    return apply(
  File "/opt/anaconda/envs/fys-stk4155/lib/python3.8/site-packages/awkward1/_util.py", line 759, in apply
    outcontent = apply(nextinputs, depth + 1)
  File "/opt/anaconda/envs/fys-stk4155/lib/python3.8/site-packages/awkward1/_util.py", line 584, in apply
    return function()
  File "/opt/anaconda/envs/fys-stk4155/lib/python3.8/site-packages/awkward1/_connect/_numpy.py", line 126, in <lambda>
    return lambda: (awkward1.layout.NumpyArray(result),)
ValueError: NumpyArray must not be scalar; try array.reshape(1)
(https://github.com/scikit-hep/awkward-1.0/blob/0.4.3/src/python/content.cpp#L1822)
```
So the higher-dimensional matrix multiplication errors out for the Awkward arrays.

I started looking at Awkward because I needed to work with ragged arrays where the two lowest dimensions would be matrices (not ragged), but it seems Awkward's support for matrix operations are lacking. Are these operations in scope for Awkward, and is the example above thus  a bug?

**Comment by jpivarski** [comment]
Not so much out of scope as not considered yet. I don't think I've overloaded whichever `__method__` corresponds to the `@`, so I don't know which method is getting called here.

Maybe there's some opportunity to do more than NumPy here: I've seen some cases where it would be handy to do matrix multiplication on many different-sized matrices in a vectorized way, using a jagged array to represent the collection of matricies. The Awkward behavior can't differ from the NumPy behavior on the same inputs, but NumPy can't take jagged arrays. Or maybe it should be a named type, like `"string"` and `"categorical"`.

**Comment by tlaundal** [comment]
> Not so much out of scope as not considered yet. I don't think I've overloaded whichever `__method__` corresponds to the `@`, so I don't know which method is getting called here.
> 
> Maybe there's some opportunity to do more than NumPy here: I've seen some cases where it would be handy to do matrix multiplication on many different-sized matrices in a vectorized way, using a jagged array to represent the collection of matricies. The Awkward behavior can't differ from the NumPy behavior on the same inputs, but NumPy can't take jagged arrays. Or maybe it should be a named type, like `"string"` and `"categorical"`.

Very good ideas. I think the proper generalization of `__matmul__`, at least when the two lowest dimensions are square, is definitely to consider the awkward arrays as collection of matrices. 

The reference docs for `ak.Array.__array_ufunc__` say "When any ufunc is applied to an Awkward Array, it applies to the innermost level of structure". I think a quick win could be to make ufuncs operate on the innermost square tensors, basically viewing all Awkward arrays as collections of tensors when the lower dimensions are square. This generalization is hopefully meaningful for other ufuncs that operate on tensors as well.

**Comment by jpivarski** [comment]
The reason this failed is because `np.matmul` is an unusual ufunc in that it operates on the last two dimensions of an array, rather than the last zero dimensions (i.e. on the scalars at the lowest level). I guess that's what makes it a gufunc. I fixed it by ensuring that any multi-node structures like `RegularArray(RegularArray( ... NumpyArray ... ))` get collapsed into a single NumpyArray node with the same shape. ("RegularArray" is an array node that makes nested equal-length lists. The only reason it exists, when NumPy's `shape` does the same job, is because it might contain something _other_ than a NumPy array.)

That solves the problem if the lists have "regular list" type. However, they might be "variable-length lists" that happen to all have the same length (i.e. ListArray or ListOffsetArray nodes). Your example doesn't touch those cases, but it's a separate case. Furthermore, they might have irregular lengths.

So I implemented the many-matrix multiplication of different sizes, as discussed above. It's the first Awkward Array feature that has been implemented using Numba, but I think that will be temporary. It can be lowered entirely into awkward-cpu-kernels.so, and therefore be part of the CPU ↔ GPU interoperability, but I don't have time for that now. In fact, writing a quick implementation is a good way to find out if anybody really needs it, to see if people use it.

The interface is:

```python
>>> import awkward1 as ak
>>> lefts = ak.Array([
...     [[1, 2],
...      [3, 4],
...      [5, 6]],
... 
...     [[1, 2, 3, 4],
...      [5, 6, 7, 8]],
... 
...     [[1],
...      [2],
...      [3],
...      [4]],
... ])
>>> rights = ak.Array([
...     [[7, 8, 9],
...      [10, 11, 12]],
... 
...     [[8, 10],
...      [11, 12],
...      [13, 14],
...      [15, 16]],
... 
...     [[5, 6, 7]],
... ])
>>> (lefts @ rights).tolist()
[
    [[ 27,  30,  33],
     [ 61,  68,  75],
     [ 95, 106, 117]],

    [[129, 140],
     [317, 348]],

    [[ 5,  6,  7],
     [10, 12, 14],
     [15, 18, 21],
     [20, 24, 28]]
]
```

but if you don't have Numba, you'll get an ImportError. Regular-length lists don't need Numba.
